src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Interop/JavaScriptExports.cs

```diff
@@ -1,270 +1,101 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Diagnostics;
-using System.Reflection;
 using System.Runtime.CompilerServices;
-using System.Threading.Tasks;
 
 namespace System.Runtime.InteropServices.JavaScript
 {
-    // this maps to src\mono\wasm\runtime\corebindings.c
-    // the methods are protected from trimming by DynamicDependency on JSFunctionBinding
+    // this maps to src\mono\wasm\runtime\corebindings.ts
+    // the methods are protected from trimming by DynamicDependency on JSFunctionBinding.BindJSFunction
     internal static unsafe partial class JavaScriptExports
     {
+
+        // The JS layer invokes this method when the JS wrapper for a JS owned object
+        //  has been collected by the JS garbage collector
         [MethodImplAttribute(MethodImplOptions.NoInlining)] // https://github.com/dotnet/runtime/issues/71425
-        public static void GetCSOwnedObjectByJSHandleRef(IntPtr jsHandle, int shouldAddInflight, out JSObject? result)
+        public static void ReleaseJSOwnedObjectByGCHandle(JSMarshalerArgument* arguments_buffer)
         {
-            lock (JSHostImplementation.s_csOwnedObjects)
+            ref JSMarshalerArgument arg_exc = ref arguments_buffer[0];
+            ref JSMarshalerArgument arg_return = ref arguments_buffer[1];
+            ref JSMarshalerArgument arg_1 = ref arguments_buffer[2];
+            try
             {
-                if (JSHostImplementation.s_csOwnedObjects.TryGetValue((int)jsHandle, out WeakReference<JSObject>? reference))
+                GCHandle handle = (GCHandle)arg_1.slot.GCHandle;
+
+                lock (JSHostImplementation.s_gcHandleFromJSOwnedObject)
                 {
-                    reference.TryGetTarget(out JSObject? jsObject);
-                    if (shouldAddInflight != 0)
-                    {
-                        jsObject?.AddInFlight();
-                    }
-                    result = jsObject;
-                    return;
+                    JSHostImplementation.s_gcHandleFromJSOwnedObject.Remove(handle.Target!);
+                    handle.Free();
                 }
+                arg_exc.Initialize();
```

Comment:
why is this initialized at the end? is it because the catch's ToJS will also implicitly initialize so you don't want to init if it failed?